src/compiler/checker.ts

```diff
@@ -29622,13 +29638,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const declaration = symbol.valueDeclaration && getRootDeclaration(symbol.valueDeclaration);
         return !!declaration && (
             isParameter(declaration) ||
-            isVariableDeclaration(declaration) && (isCatchClause(declaration.parent) || isMutableLocalVariableDeclaration(declaration))
+            isVariableDeclaration(declaration) && (isCatchClause(declaration.parent) || isMutableLocalVariableDeclaration(symbol, declaration))
         );
     }
 
-    function isMutableLocalVariableDeclaration(declaration: VariableDeclaration) {
+    function isMutableLocalVariableDeclaration(symbol: Symbol, declaration: VariableDeclaration) {
         // Return true if symbol is a non-exported and non-global `let` variable
-        return !!(declaration.parent.flags & NodeFlags.Let) && !(
+        return !!(symbol.flags & SymbolFlags.FunctionScopedVariable || (declaration.parent.flags & NodeFlags.Let)) && !(

```

Comment:
I got an answer and the reason we exclude `var` variables from being considered mutable local variables is because they have different scoping rules we would have to consider when doing the analysis we do for parameters or mutable local variables. For instance, I think `extendAssignmentPos` would have to account for an assignment that occurs before the variable declaration, since `var`s are hoisted.

So I think we can just revert this change for now and not do the new analysis for `var` variables, otherwise we'd have to possibly update every other place that calls `isParameterOrMutableLocalVariable` to account for `var`. You can leave the existing `var` tests in place though.

Reply:
reverted